blink/defbios.c

```diff
@@ -0,0 +1,180 @@
+  Put32(s->real + 0x00 * 4, kBiosSeg << 16 | (kBiosDefInt0x00 - kBiosBase));
+  Put32(s->real + 0x01 * 4, kBiosSeg << 16 | (kBiosDefInt0x01 - kBiosBase));
+  Put32(s->real + 0x02 * 4, kBiosSeg << 16 | (kBiosDefInt0x02 - kBiosBase));
+  Put32(s->real + 0x03 * 4, kBiosSeg << 16 | (kBiosDefInt0x03 - kBiosBase));
+  Put32(s->real + 0x04 * 4, kBiosSeg << 16 | (kBiosDefInt0x04 - kBiosBase));
+  Put32(s->real + 0x05 * 4, kBiosSeg << 16 | (kBiosDefInt0x05 - kBiosBase));
+  Put32(s->real + 0x06 * 4, kBiosSeg << 16 | (kBiosDefInt0x06 - kBiosBase));
+  Put32(s->real + 0x07 * 4, kBiosSeg << 16 | (kBiosDefInt0x07 - kBiosBase));
+  Put32(s->real + 0x08 * 4, kBiosSeg << 16 | (kBiosDefInt0x08 - kBiosBase));
+  Put32(s->real + 0x09 * 4, kBiosSeg << 16 | (kBiosDefInt0x09 - kBiosBase));
+  Put32(s->real + 0x0A * 4, kBiosSeg << 16 | (kBiosDefInt0x0A - kBiosBase));
+  Put32(s->real + 0x0B * 4, kBiosSeg << 16 | (kBiosDefInt0x0B - kBiosBase));
+  Put32(s->real + 0x0C * 4, kBiosSeg << 16 | (kBiosDefInt0x0C - kBiosBase));
+  Put32(s->real + 0x0D * 4, kBiosSeg << 16 | (kBiosDefInt0x0D - kBiosBase));
+  Put32(s->real + 0x0E * 4, kBiosSeg << 16 | (kBiosDefInt0x0E - kBiosBase));
+  Put32(s->real + 0x0F * 4, kBiosSeg << 16 | (kBiosDefInt0x0F - kBiosBase));
+  Put32(s->real + 0x10 * 4, kBiosSeg << 16 | (kBiosDefInt0x10 - kBiosBase));
+  Put32(s->real + 0x11 * 4, kBiosSeg << 16 | (kBiosDefInt0x11 - kBiosBase));
+  Put32(s->real + 0x12 * 4, kBiosSeg << 16 | (kBiosDefInt0x12 - kBiosBase));
+  Put32(s->real + 0x13 * 4, kBiosSeg << 16 | (kBiosDefInt0x13 - kBiosBase));
+  Put32(s->real + 0x14 * 4, kBiosSeg << 16 | (kBiosDefInt0x14 - kBiosBase));
+  Put32(s->real + 0x15 * 4, kBiosSeg << 16 | (kBiosDefInt0x15 - kBiosBase));
+  Put32(s->real + 0x16 * 4, kBiosSeg << 16 | (kBiosDefInt0x16 - kBiosBase));
+  Put32(s->real + 0x17 * 4, kBiosSeg << 16 | (kBiosDefInt0x17 - kBiosBase));
+  Put32(s->real + 0x18 * 4, kBiosSeg << 16 | (kBiosDefInt0x18 - kBiosBase));
+  Put32(s->real + 0x19 * 4, kBiosSeg << 16 | (kBiosDefInt0x19 - kBiosBase));
+  Put32(s->real + 0x1A * 4, kBiosSeg << 16 | (kBiosDefInt0x1A - kBiosBase));
+  Put32(s->real + 0x1B * 4, kBiosSeg << 16 | (kBiosDefInt0x1B - kBiosBase));
+  Put32(s->real + 0x1C * 4, kBiosSeg << 16 | (kBiosDefInt0x1C - kBiosBase));
+  Put32(s->real + 0x70 * 4, kBiosSeg << 16 | (kBiosDefInt0x70 - kBiosBase));
+  Put32(s->real + 0x71 * 4, kBiosSeg << 16 | (kBiosDefInt0x71 - kBiosBase));
+  Put32(s->real + 0x72 * 4, kBiosSeg << 16 | (kBiosDefInt0x72 - kBiosBase));
+  Put32(s->real + 0x73 * 4, kBiosSeg << 16 | (kBiosDefInt0x73 - kBiosBase));
+  Put32(s->real + 0x74 * 4, kBiosSeg << 16 | (kBiosDefInt0x74 - kBiosBase));
+  Put32(s->real + 0x75 * 4, kBiosSeg << 16 | (kBiosDefInt0x75 - kBiosBase));
+  Put32(s->real + 0x76 * 4, kBiosSeg << 16 | (kBiosDefInt0x76 - kBiosBase));
+  Put32(s->real + 0x77 * 4, kBiosSeg << 16 | (kBiosDefInt0x77 - kBiosBase));
+}
+
+void SetDefaultBiosDataArea(struct Machine *m) {
+  SetWriteAddr(m, 0x400, 0x100);
```

Comment:
The `SetWriteAddr` function is only used to highlight a single range of memory addresses changed by the last executed instruction for the TUI, and in addition only works properly when called a single time in between instructions. Thus, it seems this call and the call on Line 127 won't work as possibly intended. I'm not sure it should even be called in setup routines, since these routines are not indicative of instruction execution memory change effects.

Reply:
Yeah I don't think this does anything. Might as well remove it before we merge.

Reply:
Actually, I take it back. This highlight does appear to happen, after stepping two instructions, which is strange.

![image](https://user-images.githubusercontent.com/49262/232669242-1501a8b4-cc39-4d57-b17b-b9419aee73e3.png)

Reply:
Hello @jart, hello @ghaerr,

My main intention was to indicate that the `hvcall 0x19` was having an effect on the contents the emulated memory.

I suppose I should try invoking the `m->system->redraw` hook (either from `defbios.c` or from `blinkenlights.c`) to get Blink to properly redraw everything?

Thank you!

Reply:
> This highlight does appear to happen, after stepping two instructions, which is strange.

The highlight occurs because the `hvjmp` instruction executes `SetDefaultBiosIntVectors`, as well as `SetBiosDataArea`. However, these procedures change data in non-adjacent memory ranges.

> I suppose I should try invoking the m->system->redraw hook (either from defbios.c or from blinkenlights.c) to get Blink to properly redraw everything?

No, the problem isn't a redraw hook. The problem is that `SetWriteAddr` is called from both the above routines, and the resultant requested memory highlighting isn't cumulative, since `SetWriteAddr` only works for the last-set address range specified. Thus, the results are not likely as intended. `blink` can't at this time display multiple non-adjacent memory ranges, as it was designed more for instruction-affected memory display, rather than say, BIOS areas touched by the execution of an INT instruction.

Reply:
> Yes, for some reason the highlighting still only shows up only after I trace over the first boot sector instruction

This may be the result of some complexity in the two slightly different redraw routines, `Redraw(true)` vs `ReactiveDraw()`. I'll have to look more into this, but remember something similar while straightening this all out for `blink16`. The ReactiveDraw routine only redraws when `tuimode` is set and IIRC eventually forced some redraws within the C INT service routines by calling `Redraw(true)` directly. I can look into this further should you need help.